direct/io-blob/src/main/java/cz/o2/proxima/direct/blob/BulkBlobWriter.java

```diff
@@ -0,0 +1,175 @@
+ * limitations under the License.
+ */
+package cz.o2.proxima.direct.blob;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Preconditions;
+import com.google.common.io.ByteStreams;
+import cz.o2.proxima.annotations.Internal;
+import cz.o2.proxima.direct.bulk.AbstractBulkFileSystemAttributeWriter;
+import cz.o2.proxima.direct.bulk.FileSystem;
+import cz.o2.proxima.direct.bulk.Path;
+import cz.o2.proxima.direct.core.BulkAttributeWriter;
+import cz.o2.proxima.direct.core.CommitCallback;
+import cz.o2.proxima.direct.core.Context;
+import cz.o2.proxima.repository.EntityDescriptor;
+import cz.o2.proxima.storage.StreamElement;
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.net.URI;
+import lombok.AccessLevel;
+import lombok.Getter;
+import lombok.extern.slf4j.Slf4j;
+
+/** {@link BulkAttributeWriter} for blob storages. */
+@Internal
+@Slf4j
+public abstract class BulkBlobWriter<BlobT extends BlobBase, AccessorT extends BlobStorageAccessor>
+    implements BulkAttributeWriter {
+
+  @Getter private final AccessorT accessor;
+  @Getter private final Context context;
+  private transient BulkAttributeWriter wrap;
+
+  @Getter(AccessLevel.PACKAGE)
+  private transient boolean initialized;
+
+  public BulkBlobWriter(AccessorT accessor, Context context) {
+    this.accessor = accessor;
+    this.context = context;
+    init();
+  }
+
+  /** Retrieve {@link EntityDescriptor} of this {@link BulkAttributeWriter}. */
+  public EntityDescriptor getEntityDescriptor() {
+    return accessor.getEntityDescriptor();
+  }
+
+  @Override
+  public URI getUri() {
+    return accessor.getUri();
+  }
+
+  @Override
+  public void write(StreamElement data, long watermark, CommitCallback statusCallback) {
+    init();
+    wrap.write(data, watermark, statusCallback);
+  }
+
+  @Override
+  public void updateWatermark(long watermark) {
+    wrap.updateWatermark(watermark);
+  }
+
+  @Override
+  public void rollback() {
+    wrap.rollback();
+  }
+
+  private void init() {
+    if (!initialized) {
+      File tmpDir = accessor.getTmpDir();
+      FileSystem localFs = FileSystem.local(tmpDir, accessor.getNamingConvention());
+      wrap =
+          new AbstractBulkFileSystemAttributeWriter(
+              getEntityDescriptor(),
+              accessor.getUri(),
+              localFs,
+              accessor.getNamingConvention(),
+              accessor.getFileFormat(),
+              context,
+              accessor.getRollPeriod(),
+              accessor.getAllowedLateness()) {
+
+            @Override
+            protected void flush(Bulk bulk) {
+              BulkBlobWriter.this.flush(bulk.getPath(), bulk.getMaxTs());
+            }
+          };
+
+      if (!tmpDir.exists()) {
+        tmpDir.mkdirs();
+      } else if (tmpDir.isDirectory()) {
+        removeDir(tmpDir);
+        tmpDir.mkdirs();
```

Comment:
this seems incorect, what happens if we have multiple instances sharing the fs?

Reply:
I think this needs fixing in `BlobStorageAccessor.getTmpDir()`. Fixed so that the directory there is always unique.